### ISSUE TITLE
Search engines in horizontal card with only their icons

### ISSUE THREAD

**Issue by magnus-ISU** [TBog/TBLauncher#339]
**Description**

Right now, I added a bunch of custom search engines so its pretty annoying to open the one I actually want up as they are so high on the screen.

I also added icons to all of them.

Inspired by StarlightLauncher from r/android, I think it would make sense to show only the icons for search engines, since they don't change from search to search so the extra information of the app name is not useful.

Here's a screenshot of the current situation:
![image](https://user-images.githubusercontent.com/75949275/184983605-ef213b81-bef3-4b1d-9a0e-eff573b60910.png)

**Solution**

Search engines show at the top as currently but in rows of icons.

![image](https://user-images.githubusercontent.com/75949275/184985289-c7dc38e9-b6ff-44f9-a932-f41f3188e26a.png)

**Comment by TBog** [comment]
I like this idea, although it will work only if you have different icons for the search engines.

**Comment by magnus-ISU** [comment]
To get around that you could just show blank icons as holding the list entry now allows you to change an icon and so upon seeing a blank icon you would probably be motivated to change it.

**Comment by ltGuillaume** [comment]
When changing an icon, TBLauncher already suggests an icon based on the app name's "initials". I guess that could be used for search engine default icons in such a layout, too.

**Comment by TBog** [comment]
The difficult part is the layout for this. Right now I can display the results in a list or in a grid.

I see two solutions:
1. The result list LayoutManager would display the search engines in a grid while displaying the other results in a list.
2. Have two result list views. One for the search results and one for the search engines. But when and how do I transition from one to the other?

**Comment by magnus-ISU** [comment]
I think what would make the most sense, knowing very little about android development, is to have a list and a grid, the list below the grid and taking up the minimum possible space, and the grid taking up whatever space it must. The grid is only shown if something is typed into the search bar. The 'transition'? is the same as whatever separates the dock from the results list.

You also probably want this to be a toggled behaviour to show only icons - you'd have to decide whether when in list mode (as opposed to icons) current behaviour would be preserved or a spacer would exist between apps and search engines still, as I think that would be easier to implement and maybe also better.

**Comment by daffydock** [comment]
Number 2 seems like a good option. That way, people could decide how they want the search engines be displayed. 

Mind you, my focus would be ease of reach. I too have a inner of search engines and a gridakes them easier to reach over a list. Labels or no labels.

**Comment by TBog** [comment]
> The grid is only shown if something is typed into the search bar.

This way the search engines will be visible immediately.
I am just worried that it will take screen space real-estate when you are searching for something else.
I was thinking about the transition thinking that I should show the search engine grid only after scrolling the result list to the end.

I can implement this fairly quickly and make a test build so you can test and give feedback.

**Comment by TBog** [comment]
If you have time please try out the [test build](https://github.com/TBog/TBLauncher/releases/tag/v7.1-search_engine_grid.1) and provide some feedback.

**Comment by magnus-ISU** [comment]
Right now its putting 8 in a row, that seems like too much but I assume will be configurable in the future.

My icons didn't work somehow, or weren't remembered. I can still change them again, but you probably don't want that to mess up for an official release.

The list appears only at the top of the screen, even if there are only a couple results. Personally, I imagined it working as you say, only showing search engines after scrolling other results to the end. Also, if there are only a few results, showing the search engines right above them rather than still at the top of the screen.

![Screenshot_2022-08-18_10-02-35](https://user-images.githubusercontent.com/75949275/185430470-6762b861-5bad-4b79-95a6-18c8834c0b51.png)

There is currently a thin line (an empty grid) at the top of the screen when nothing is typed, the above should fix that. The grid should not be visible at all if nothing is typed.

![Screenshot_2022-08-18_10-02-23](https://user-images.githubusercontent.com/75949275/185430692-489eaffb-5d80-4fdb-8a7a-42400036a6f8.png)

If you type in something that matches no apps, contacts, etc the grid disappears too and you can't search at all. For some reason unable to take a screenshot of this without resetting the text field.

If possible, I'd reverse the order you draw the icons. The bottommost things are easiest to click.

I like it so far though.

**Comment by TBog** [comment]
> Right now its putting 8 in a row, that seems like too much but I assume will be configurable in the future.

If you change the icon size (possibly needing a launcher restart) the number of icons will change. This is just a proof of concept for the second solution I mentioned.

---

> Personally, I imagined it working as you say, only showing search engines after scrolling other results to the end.

I was trying to implement what I understood from your description.
`
I think what would make the most sense, knowing very little about android development, is to have a list and a grid, the list below the grid and taking up the minimum possible space, and the grid taking up whatever space it must. The grid is only shown if something is typed into the search bar. The 'transition'? is the same as whatever separates the dock from the results list.
`
I did not change the result list height because that is linked to the keyboard hiding when scrolling animation.

---

My preferred way to show the search engines would be to group them in the result list, at the end, just like we have them in the stable release. By group I mean have a result list row that can hold multiple icons.

**Comment by magnus-ISU** [comment]
Yeah, I guess that makes sense if its too hard to get the result list to work with the scrolling thing, which thinking about it I'm not sure how you could handle.

I agree that search engines should be on screen only when there are few results, maybe I didn't communicate that well.

I just liked the idea of separating the results groups entirely, like how the dock and results list are separate, from a UI point of view. And having the result list take up the minimum amount of space, as in the first UI mockup. But you're probably right it would be tough to achieve that with the scrolling animation.

**Comment by daffydock** [comment]
Tried the test build. I get 9 icons --that is how I have set it. So no worries about that. I too get the search icons resetting to a number of magnifying glass icons. Away from the ones set. Even after restarting the app. 

Is there a way to have that separate group appear at the bottom of the screen? On top of the search bar/dock, rather than at the very top? Simply because they would be far easier ro reach that way. Like the general idea. 

I thought of getting the results in a grid, but this is fairly okay, too. Thanks, so far.

**Comment by daffydock** [comment]
Hi @TBog, hope you do mind me asking. I was curious about your ongoing thoughts about this idea. I agree that it seems like a very good addition and was wondering about your take, as we have not heard much on it after testing that APK back in August. No rush on responding right away. Not going anywhere. Cheers.

**Comment by TBog** [comment]
I was just very busy. Just had some time to fix the preferences / settings UI. I'll get back to the horizontal cards this year.
